itest/send_test.go

```diff
@@ -951,6 +951,259 @@ func testReattemptFailedSendUniCourier(t *harnessTest) {
+			expectedEventCount,
+		)
+	}()
+
+	// Start asset transfer and then mine to confirm the associated on-chain
+	// tx. The on-chain tx should be mined successfully, but we expect the
+	// asset proof transfer to be unsuccessful.
+	sendAssetsToAddr(t, sendTapd, recvAddr)
+	MineBlocks(t.t, t.lndHarness.Miner.Client, 1, 1)
+
+	// There may be a delay between mining the anchoring transaction and
+	// recognizing its on-chain confirmation. To handle this potential
+	// delay, we use require.Eventually to ensure the transfer details are
+	// correctly listed after confirmation.
+	require.Eventually(t.t, func() bool {
+		// Ensure that the transaction took place as expected.
+		listTransfersResp, err := sendTapd.ListTransfers(
+			ctxb, &taprpc.ListTransfersRequest{},
+		)
+		require.NoError(t.t, err)
+
+		require.Len(t.t, listTransfersResp.Transfers, 1)
+
+		firstTransfer := listTransfersResp.Transfers[0]
+		require.NotEqual(t.t, firstTransfer.AnchorTxHeightHint, 0)
+		require.NotEmpty(t.t, firstTransfer.AnchorTxBlockHash)
+
+		// Assert proof transfer status for each transfer output.
+		require.Len(t.t, firstTransfer.Outputs, 2)
+
+		// First output should have a proof delivery status of not
+		// applicable. This indicates that a proof will not be delivered
+		// for this output.
+		firstOutput := firstTransfer.Outputs[0]
+		require.Equal(
+			t.t, taprpc.ProofDeliveryStatusNotApplicable,
+			firstOutput.ProofDeliveryStatus,
+		)
+
+		// The second output should have a proof delivery status of
+		// pending. This indicates that the proof deliver has not yet
+		// completed successfully.
+		secondOutput := firstTransfer.Outputs[1]
+		require.Equal(
+			t.t, taprpc.ProofDeliveryStatusPending,
+			secondOutput.ProofDeliveryStatus,
+		)
+
+		return true
+	}, defaultWaitTimeout, 200*time.Millisecond)
+
+	// Wait to ensure that the asset transfer proof deliver attempt has been
+	// made.
+	wg.Wait()
+
+	// Attempt to send the change output to the receiver node. This
+	// operation should select the change output from the previous
+	// transaction and transmit it to the receiver node, despite the fact
+	// that proof delivery for the previous transaction remains incomplete
+	// (due to the proof courier being shut down). We will generate a new
+	// address for this new transaction.
+	recvAddr, err = recvTapd.NewAddr(ctxb, &taprpc.NewAddrRequest{
+		AssetId:          genInfo.AssetId,
+		Amt:              42,
+		ProofCourierAddr: proofCourierAddr,
+	})
+	require.NoError(t.t, err)
+	AssertAddrCreated(t.t, recvTapd, rpcAssets[0], recvAddr)
+
+	sendAssetsToAddr(t, sendTapd, recvAddr)
+	MineBlocks(t.t, t.lndHarness.Miner.Client, 1, 1)
+
+	// There may be a delay between mining the anchoring transaction and
+	// recognizing its on-chain confirmation. To handle this potential
+	// delay, we use require.Eventually to ensure the transfer details are
+	// correctly listed after confirmation.
+	require.Eventually(t.t, func() bool {
```

Comment:
nit: this could be pulled into the `assertions` file if we're going to use it across multiple tests.

non-blocking / could be in the follow-up PR.

Reply:
I'm not sure if I agree with you that this should be pulled into `assertions`. All I'm doing here is calling the `ListTransfers` endpoint and then inspecting the results. And the results are confirmed anchor transactions. And for each transaction, one not applicable proof delivery status output and one delivery pending output.

Is that really a repeatable state that we can package up into an assertion? What would we even name that assertion?

Reply:
We have a similar helper func that inspects the output of `ListTransfers`:

https://github.com/lightninglabs/taproot-assets/blob/39c1f2786bd54526165a671078663b8ddfba913b/itest/assertions.go#L1060

https://github.com/lightninglabs/taproot-assets/blob/39c1f2786bd54526165a671078663b8ddfba913b/itest/assertions.go#L1095

The more flexible version we use for inspecting slices of assets:

https://github.com/lightninglabs/taproot-assets/blob/39c1f2786bd54526165a671078663b8ddfba913b/itest/assertions.go#L276

Which is called as part of larger assertions or helper funcs.

Reply:
OK, thanks for the example. I'll come back to this in a future PR.